### DISCUSSION TITLE
Finally! Automated Authentication Made Easy with Python

### DISCUSSION THREAD

**Discussion by slider1234** [etsy/open-api#1236] [Show and tell]
[GetNewToken.zip](https://github.com/user-attachments/files/15886102/GetNewToken.zip)
I've updated the Download on 6/18/24 
It was causing a loop trying to authenticate. A simple pause was added to allow the .env file time to update while running and it's fixed now.

I attached a zip file here as an attachment that has a GetNewToken.py file, an .env file and StartAuthenticator.bat. All three files must be in the same folder/directory. This is your replacement for the horrible 'Quick Start Tutorial' and because this one is completely automated with no knowledge of coding or struggling needed.

It doesn't make you jump through hoops as the tutorial has you do. 
- There are no folders to create. 
- There are no HBS files to make. 
- You don't have to copy and paste anything any longer.

To get this scripts working, you'll need to do a few things in the .env file first. This file is where important information is stored to help the scripts connect to your Etsy account. There are more direct instructions within the .env file to help you if you've never done this before.

1. Getting Your API Key:

- CLIENT_ID: You'll need an API key, which is a unique identifier for your Etsy app. You can get one by going to the Etsy developer portal and creating a new app.

2. Finding Your Shop ID:

- ETSY_SHOP_ID: Each Etsy shop has a unique identifier called the Shop ID. You can find yours by viewing the HTML source of your shop’s page or contacting Etsy support.

3. Deciding What to Do After Login:

- RUN_SCRIPT_AFTER_AUTH: Decide if you want to run another script after logging in. Set this to 'true' to automate tasks like updating listings.
- RETURN_TO_PROCESSES: Enter the name of the script file that will run after logging in, like ProcessUploadListings.py.
Callback URL:

- REDIRECT_URI: Use http://localhost:3003/oauth/redirect and update the Callback URL in your Etsy app settings to match.

Here is what you don't have to do any longer: These fill in within the .env file automatically.
CODE_CHALLENGE=
CLIENT_VERIFIER=
ETSY_OAUTH_TOKEN=
ETSY_OAUTH_TOKEN_EXPIRY=
STATE_ID=

You will need to start the GetNewToken.py file with the StartAuthenticator.bat file, so you can see the output of what is happening.
That's it!! 

At any point in code  you make yourself for creating, updating or removing listings and your token expires you can have your code call GetNewToken.py and then have a return to the file it left by adding to the .env file where RETURN_TO_PROCESSES is located, so there is nothing to do manually to slow you down.

**Comment by DevTKSS** [comment]
Maybe you would like to add, that the way over the browsers dev.tools panel maybe show the text "shop_id" but will NOT result in the number we do search. I tryed it this way because I thought it would be the same (yes, you did discribe it absolutly right!)...

The values will only be viewable throught the raw source code where you only find this never ending html code XD
- right click
- Inspect
- way down (in chrome in my case) is a panel named "search" and there you type in the "shop_id" and et voila 👍 the searched number did appear

**Reply by slider1234** [reply]
If you've tried it out then be sure to let me know what you think of it.

**Reply by DevTKSS** [reply]
@slider1234 Hey there,
Had to learn a whole new coding language (c#) and now I am trying again to call the Etsy API again, but I am struggling with the auth Endpoint again. Readed your sample and now they are having the me Endpoint added but I am confused about the query and header parameters ...
As asked in #1443 
Do we really send our api-key aka keystring in the bare query? Or did I miss something? I mean I don't know much so far about http but normally you don't send credentials over the network so why there?

**Comment by DevTKSS** [comment]
tryed the "GitHub Copilot", lets see if the AI works so well like they are saying.
Since I will write my application for angular, so in typescript, I asked it to translate it for me :) I will try it out and if it works I would post it for the others here who like to use it

**Comment by slider1234** [comment]
I've updated the Download on 6/18/24 
It was causing a loop trying to authenticate. A simple pause was added to allow the .env file time to update while running and it's fixed now.
You can download the file from the main post.
